modules/system/defaults/dock.nix

```diff
@@ -128,18 +128,75 @@ in {
     };
 
     system.defaults.dock.persistent-apps = mkOption {
-      type = types.nullOr (types.listOf (types.either types.path types.str));
+      type = let
+        taggedType = types.attrTag {
+              app = mkOption {
+                description = "An application to be added to the dock.";
+                default = null;
+                type = types.submodule {
+                  options.path = mkOption {
+                    description = "Path to the application.";
+                    type = types.str;
+                  };
+                };
+              };
+              spacer = mkOption {
+                description = "A spacer to be added to the dock. Can be small or regular size.";
+                default = {
+                  small = false;
+                };
+                type = types.submodule {
+                  options.small = mkOption {
+                    description = "Whether the spacer is small.";
+                    type = types.bool;
+                  };
+                };
+              };
+              folder = mkOption {
+                description = "A folder to be added to the dock.";
+                default = null;
+                type = types.submodule {
+                  options.path = mkOption {
+                    description = "Path to the folder.";
+                    type = types.str;
+                  };
+                };
+              };
+            };
+        in
+      types.nullOr (types.listOf (types.coercedTo (types.either types.str types.path)(path: {app = {inherit path;};}) taggedType));
       default = null;
-      example = [ "/Applications/Safari.app" "/System/Applications/Utilities/Terminal.app" ];
+      example = [
+        { app = { path = "/Applications/Safari.app"; }; }
+        { spacer = { small = false; }; }
+        { spacer = { small = true; }; }
+        { folder = { path = "/System/Applications/Utilities"; }; }
+      ];
       description = ''
-        Persistent applications in the dock.
+        Persistent applications, spacers, and folders in the dock.
       '';
-      apply = value:
-        if !(isList value)
-        then value
-        else map (app: { tile-data = { file-data = { _CFURLString = app; _CFURLStringType = 0; }; }; }) value;
+      apply =
+      let
+        toTile = item: if item ? app then {
+          tile-data = { file-data = { _CFURLString = item.app.path; _CFURLStringType = 0; }; };
```

Comment:
Maybe I've missed some nix-darwin style preference, but IMO this is more readable as:

```nix
tile-data.file-data = {
  _CFURLString = item.app.path;
  _CFURLStringType = 0;
};
```

Reply:
I haven't seen any formatter configuration or documentation on formatting in this repo. So, I have no idea whether I should just format the whole file with nixfmt or just try to fit it into whats in here...

Reply:
There was a discussion in #1054 that would be worth revisiting. I agree that a default format would be really helpful.